applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_local_stress_response_function.cpp

```diff
@@ -328,5 +343,185 @@ namespace Kratos
+
+        for(IndexType gp_it = 0; gp_it < num_GP; ++gp_it)
+        {
+            for(IndexType i = 0; i < rElementalDofList.size(); ++i)
+            {
+                if (rElementalDofList[i]->GetVariable().Name() == rDofLabel)
+                {
+                    if (rElementalDofList[i]->Id() == id_node_1)
+                        rResult[i] += prefactor * (num_GP - gp_it);
+                    else if (rElementalDofList[i]->Id() == id_node_2)
+                        rResult[i] += -prefactor * (gp_it + 1);
+                }
+            }
+        }
+        rResult /= num_GP;
+
+        KRATOS_CATCH("");
+    }
+
+    void AdjointLocalStressResponseFunction::CalculateGPParticularSolutionLinearElement2N(Vector& rResult,
+                            DofsVectorType &rElementalDofList, const std::string& rDofLabel) const
+    {
+        KRATOS_TRY;
+
+        if(rResult.size() != rElementalDofList.size())
+            rResult.resize(rElementalDofList.size());
+
+        const unsigned int num_GP = mpTracedElement->GetGeometry().IntegrationPointsNumber(mpTracedElement->GetIntegrationMethod());
+        const double prefactor = 1.0 / (1.0 + num_GP);
+
+        const IndexType id_node_1 = mpTracedElement->GetGeometry()[0].Id();
+        const IndexType id_node_2 = mpTracedElement->GetGeometry()[1].Id();
+
+        for(IndexType i = 0; i < rElementalDofList.size(); ++i)
+        {
+            if (rElementalDofList[i]->GetVariable().Name() == rDofLabel)
+            {
+                if (rElementalDofList[i]->Id() == id_node_1)
+                    rResult[i] = prefactor * (num_GP + 1 - mIdOfLocation);
+                else if (rElementalDofList[i]->Id() == id_node_2)
+                    rResult[i] = -prefactor * mIdOfLocation;
+            }
+        }
+
+        KRATOS_CATCH("");
+    }
+
+    void AdjointLocalStressResponseFunction::CalculateNodeParticularSolutionLinearElement2N(Vector& rResult,
+                     DofsVectorType &rElementalDofList, const std::string& rDofLabel) const
+    {
+        KRATOS_TRY;
+
+        if(rResult.size() != rElementalDofList.size())
+            rResult.resize(rElementalDofList.size());
+
+        for(IndexType i = 0; i < rElementalDofList.size(); ++i)
+        {
+            if (rElementalDofList[i]->Id() == mpTracedElement->GetGeometry()[mIdOfLocation-1].Id() &&
+                rElementalDofList[i]->GetVariable().Name() == rDofLabel)
+            {
+                if (mIdOfLocation == 1)
+                    rResult[i] = 1.0;
+                else if (mIdOfLocation == 2)
+                    rResult[i] = -1.0;
+            }
+        }
+
+        KRATOS_CATCH("");
+    }
+
+    void AdjointLocalStressResponseFunction::FindCorrespondingDofLabel(std::string& rDofLabel) const
+    {
+        KRATOS_TRY;
+
+        switch (mTracedStressType)
+        {
+            case TracedStressType::MX:
+            {
+                rDofLabel = "ADJOINT_ROTATION_X";
```

Comment:
I have the feeling that passing a variable instead of a string would be more efficient here. up to the point where you compare with the dof variable in CalculateNodeParticularSolutionLinearElement2N

Reply:
then the name for this function would be FindVariableComponent or similar

Reply:
In that case a typdef for variable component would be good i guess =)